browser/ui/views/brave_actions/brave_news_action_view.cc

```diff
@@ -0,0 +1,200 @@
+        ->RemoveObserver(this);
+  }
+}
+
+void BraveNewsActionView::Init() {
+  Update();
+}
+
+void BraveNewsActionView::Update() {
+  if (!should_show_.GetValue()) {
+    SetVisible(false);
+    return;
+  }
+
+  auto* contents = tab_strip_->GetActiveWebContents();
+  bool subscribed = false;
+  absl::optional<BraveNewsTabHelper::FeedDetails> feed = absl::nullopt;
+
+  if (contents) {
+    auto* tab_helper = BraveNewsTabHelper::FromWebContents(contents);
+    if (!tab_helper->GetAvailableFeeds().empty()) {
+      feed = tab_helper->GetAvailableFeeds()[0];
+      subscribed = tab_helper->IsSubscribed();
+    }
+  }
+
+  auto background =
+      subscribed ? views::CreateRoundedRectBackground(
+                       selectedColor,
+                       ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
+                           views::Emphasis::kMaximum, GetPreferredSize()))
+                 : nullptr;
+  auto image =
+      gfx::CreateVectorIcon(kBraveTodaySubscribeIcon, 16,
+                            color_utils::DeriveDefaultIconColor(
+                                subscribed ? SK_ColorWHITE : SK_ColorBLACK));
+  SetImage(ButtonState::STATE_NORMAL, image);
+  SetBackground(std::move(background));
+  SetVisible(!!feed);
+}
+
+SkPath BraveNewsActionView::GetHighlightPath() const {
+  auto highlight_insets = gfx::Insets();
+  gfx::Rect rect(GetPreferredSize());
+  rect.Inset(highlight_insets);
+  const int radii = ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
+      views::Emphasis::kMaximum, rect.size());
+  SkPath path;
+  path.addRoundRect(gfx::RectToSkRect(rect), radii, radii);
+  return path;
+}
+
+std::u16string BraveNewsActionView::GetTooltipText(const gfx::Point& p) const {
+  return l10n_util::GetStringUTF16(IDS_BRAVE_NEWS_ACTION_VIEW_TOOLTIP);
+}
+
+std::unique_ptr<views::LabelButtonBorder>
+BraveNewsActionView::CreateDefaultBorder() const {
+  std::unique_ptr<views::LabelButtonBorder> border =
+      LabelButton::CreateDefaultBorder();
+  border->set_insets(gfx::Insets::TLBR(3, 0, 3, 0));
+  return border;
+}
+
+void BraveNewsActionView::OnTabStripModelChanged(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange& change,
+    const TabStripSelectionChange& selection) {
+  if (selection.active_tab_changed()) {
+    if (selection.old_contents) {
+      BraveNewsTabHelper::FromWebContents(selection.old_contents)
+          ->RemoveObserver(this);
+    }
+
+    if (selection.new_contents) {
+      BraveNewsTabHelper::FromWebContents(selection.new_contents)
+          ->AddObserver(this);
+    }
+  }
+
+  Update();
+}
+
+void BraveNewsActionView::OnAvailableFeedsChanged(
+    const std::vector<BraveNewsTabHelper::FeedDetails>& feeds) {
+  Update();
+}
+
+void BraveNewsActionView::ButtonPressed() {
+  // If the bubble is opened, closed it.
+  if (bubble_widget_) {
+    bubble_widget_->Close();
+    return;
+  }
+
+  if (!tab_strip_->GetActiveWebContents())
+    return;
+
+  auto* tab_helper =
```

Comment:
Idea: Or we can control this button's enabled state based on available feeds state.
That means, button pressed callback is not called in disabled state.

Reply:
So I had a think about it, and I'm pretty sure this wasn't actually needed because I'm already hiding the button if there are no feeds. I've removed it completely :smile:

Reply:
Great :)